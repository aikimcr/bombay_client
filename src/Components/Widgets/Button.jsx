import React from 'react';
import PropTypes from 'prop-types';

function Button(props) {
  const classList = new Set();

  if (props.className) {
    const propsClasses = props.className.split(/\s+/);
    propsClasses.forEach((name) => classList.add(name));
  }

  classList.add('btn');
  const className = Array.from(classList).join(' ');

  if (props.disabled) {
    return (
      <button className={className} onClick={props.onClick} disabled>
        {' '}
        {props.label}
      </button>
    );
  } else {
    return (
      <button className={className} onClick={props.onClick}>
        {props.label}
      </button>
    );
  }
}

Button.propTypes = {
  label: PropTypes.string.isRequired,
  disabled: PropTypes.bool.isRequired,
  onClick: PropTypes.func.isRequired,
  className: PropTypes.string,
};

export default Button;
