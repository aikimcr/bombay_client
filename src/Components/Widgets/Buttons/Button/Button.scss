.button {
  color: var(--primary-button-color);
  background-color: var(--primary-button-background-color);
  border-radius: 2rem;
  border-style: ridge;
  border-width: .25em;
  border-color: var(--primary-button-background-color);
  font-size: var(--standard-button-font-size);
  line-height: 1;
  padding: .25em .5em;
  cursor: pointer;

  &__secondary {
    color: var(--secondary-button-color);
    background-color: var(--secondary-button-background-color);
    border-color: var(--secondary-button-background-color);
  }

  &:hover:not(:disabled) {
    border-style: outset;
  }

  &:active:not(:disabled) {
    border-style: inset;
  }

  &:disabled {
    cursor: default;
    color: rgb(50%, 50%, 50%);
    opacity: .5;
  }
}