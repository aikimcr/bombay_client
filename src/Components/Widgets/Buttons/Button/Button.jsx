import React from "react";

import "./Button.scss";
import classnames from "classnames";

export const Button = ({
  className,
  text,
  role = "primary",
  onClick,
  ...rest
}) => {
  const classNames = classnames([className, "button"], {
    button__primary: role === "primary",
    button__secondary: role === "secondary",
  });
  return (
    <button {...rest} className={classNames} onClick={onClick}>
      {text}
    </button>
  );
};
