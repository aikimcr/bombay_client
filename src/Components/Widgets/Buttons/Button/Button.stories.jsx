import { fn } from 'storybook/internal/test';
import { Button } from './Button';

const meta = {
  component: Button,
  args: {
    className: 'storybook-button',
    text: 'A Button',
    onClick: fn(),
  },
};

export default meta;

export const Default = {};

export const Secondary = {
  args: {
    role: 'secondary',
  },
};
