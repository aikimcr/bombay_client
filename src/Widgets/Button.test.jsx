import { act, render, screen } from "@testing-library/react";

import Button from "./Button.jsx";

jest.useFakeTimers();

it("should show a simple button", async () => {
  const clickHandler = jest.fn();
  const disabled = false;

  const { asFragment } = render(
    <Button label="xyzzy" disabled={disabled} onClick={clickHandler} />,
  );

  expect(asFragment).toMatchSnapshot();
});

it("should respond to click", async () => {
  const clickHandler = jest.fn();
  const disabled = false;

  render(<Button label="xyzzy" disabled={disabled} onClick={clickHandler} />);
  const actualButton = screen.getByText(/xyzzy/);
  actualButton.click();

  expect(clickHandler).toBeCalled();
});

it("should ignore the click", async () => {
  const clickHandler = jest.fn();
  const disabled = true;

  render(<Button label="xyzzy" disabled={disabled} onClick={clickHandler} />);
  const actualButton = screen.getByText(/xyzzy/);
  actualButton.click();

  expect(clickHandler).not.toBeCalled();
});

it("should respect classnames", async () => {
  const clickHandler = jest.fn();
  const disabled = true;

  render(
    <Button
      className="foobar"
      label="xyzzy"
      disabled={disabled}
      onClick={clickHandler}
    />,
  );
  const actualButton = screen.getByText(/xyzzy/);
  expect(actualButton).toHaveClass("btn");
  expect(actualButton).toHaveClass("foobar");
});
