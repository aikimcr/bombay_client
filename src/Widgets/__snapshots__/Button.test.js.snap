// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should show a simple button 1`] = `[Function]`;
